Code/Framework/AzCore/AzCore/Memory/SystemAllocator.cpp

```diff
@@ -17,15 +17,23 @@
 
 #include <AzCore/Debug/Profiler.h>
 
-#define AZCORE_SYS_ALLOCATOR_HPPA  // If you disable this make sure you start building the heapschema.cpp
-//#define AZCORE_SYS_ALLOCATOR_MALLOC
+#define AZCORE_SYSTEM_ALLOCATOR_HPHA 1
+#define AZCORE_SYSTEM_ALLOCATOR_MALLOC 2
+#define AZCORE_SYSTEM_ALLOCATOR_HEAP 3
 
-#ifdef AZCORE_SYS_ALLOCATOR_HPPA
-#   include <AzCore/Memory/HphaSchema.h>
-#elif defined(AZCORE_SYS_ALLOCATOR_MALLOC)
-#include <AzCore/Memory/MallocSchema.h>
+#if !defined(AZCORE_SYSTEM_ALLOCATOR)
+    // define the default
+    #define AZCORE_SYSTEM_ALLOCATOR AZCORE_SYSTEM_ALLOCATOR_HPHA
+#endif
+
+#if AZCORE_SYSTEM_ALLOCATOR == AZCORE_SYSTEM_ALLOCATOR_HPHA
+    #include <AzCore/Memory/HphaSchema.h>
+#elif AZCORE_SYSTEM_ALLOCATOR == AZCORE_SYSTEM_ALLOCATOR_MALLOC
+    #include <AzCore/Memory/MallocSchema.h>
+#elif AZCORE_SYSTEM_ALLOCATOR == AZCORE_SYSTEM_ALLOCATOR_HEAP
+    #include <AzCore/Memory/HeapSchema.h>
 #else
-#   include <AzCore/Memory/HeapSchema.h>
+    #error "Invalid allocator selected for SystemAllocator"
 #endif
```

Comment:
the HphaSchema doesnt play well with ASan because most of the memory is not returned. The hooks to ASan happen at the OS allocation level. Changed this to be able to pick the schema to use for the SystemAllocator (when ASan is enabled it will use `MallocSchema`)

Reply:
Did you get a chance to look at the asan interface? I think you can unpoison regions or mark regions as being freed by some other mechanism.

https://github.com/llvm-mirror/compiler-rt/blob/master/include/sanitizer/asan_interface.h

Reply:
that is clang, MSVC doesnt have that interface yet and this implementation is just for MSVC